test/AutoDiff/compiler_crashers_fixed/issue-66522-pullback-generation-when-tangentvector-of-input-contains-nondifferentiable-fields.swift

```diff
@@ -0,0 +1,67 @@
+// RUN: %target-swift-frontend -emit-sil -O %s
+
+import _Differentiation
+
+// Issue #66522:
+// Pullback generation for a function mapping a differentiable input type to 
+// one of its differentiable fields fails when the input's tangent vector contains
+// non-differentiable fields.                          
+public struct P<Value>: Differentiable
+where
+    Value:  Differentiable,
+    Value.TangentVector == Value,
+    Value: AdditiveArithmetic {
+  // `P` is its own `TangentVector`
+  public typealias TangentVector = Self
+  
+  // Non-differentiable field in `P`'s `TangentVector`.
+  public let name: String = ""
+  var value: Value
+}
+
+extension P: Equatable, AdditiveArithmetic
+where Value: AdditiveArithmetic {
+    public static var zero: Self {fatalError()}
+    public static func + (lhs: Self, rhs: Self) -> Self {fatalError()}
+    public static func - (lhs: Self, rhs: Self) -> Self {fatalError()}
+}
+
+@differentiable(reverse)
+internal func testFunction(data: P<Double>) -> Double {
+    data.value
+}
+
+
+// WHAT?
+// This test case currently fails with the following compiler error -
+// ```
+// /Users/kshitij/workspace/swift-project/swift/test/AutoDiff/compiler_crashers_fixed/issue-66522-pullback-generation-when-tangentvector-of-input-contains-nondifferentiable-fields.swift:9:15: error: expression is not differentiable
+// public struct P<Value>: Differentiable
+//               ^
+// /Users/kshitij/workspace/swift-project/swift/test/AutoDiff/compiler_crashers_fixed/issue-66522-pullback-generation-when-tangentvector-of-input-contains-nondifferentiable-fields.swift:9:15: note: cannot differentiate access to property 'P.name' because property type 'String' does not conform to 'Differentiable'
+// public struct P<Value>: Differentiable
+//               ^
+// ```
+// 
+// WHY?
+// The `TangentVector` for `P` has been defined as `Self`.
+// The pullback generation for `P.init` fails because it now 
+// contains a projection to a non-differentiable stored
+// property, which is not allowed currently - https://github.com/apple/swift/blob/main/lib/SILOptimizer/Differentiation/PullbackCloner.cpp#L1850.
+//
+// The same issue does not happen if we let the compiler
+// generated `P`'s `TangentVector`, because `P.init` now 
+// becomes differentiable. 
+// 
+// SO WHAT? 
+// It feels like a bit of an edge case that for a user-defined tangent 
+// vector, such as that for `P`, we cannot have a function mapping 
+// individual member fields to `P`.
+//
+// NOW WHAT?
+// Is this edge case something that we need to handle?
+// 
+// @differentiable(reverse)
```

Comment:
@rxwei Could you please take a look at the comments on this commented-out differentiable function? 

I may have discovered an edge-case and I wanted to confirm whether or not it's something we need to handle and/or fix.